### PR TITLE
Fix race condition in product check

### LINKED ISSUE #8268: Exception thrown trying to await requests tasks after multiple SearchAsync are called
**Elastic.Clients.Elasticsearch version**: 8.14.5

**Elasticsearch version**: 8.7.1

**.NET runtime version**: .NET 6

**Operating system version**: Windows 11

**Description of the problem including expected versus actual behavior**:
Calling multiple times SearchAsync and awaiting the responses after results in "Object reference not set to an instance of an object."

**Steps to reproduce**:
 1. Initialize a SearchRequest
 2. Call multiple times SearchAsync with the same SearchRequest and do NOT await the response, get the tasks instead
 3. Await the tasks execution after

**Expected behavior**
The tasks should be awaited correctly, returning the results of the searches.

**Provide `DebugInformation` (if relevant)**:
This wasn't an issue with NEST and ES7, it was reproduced after trying to upgrade to Elastic.Clients.Elasticsearch and ES8. 
It was reproduced initially by running some tests that worked before the updates, that called some SearchAsync and CountAsync consecutively. **The exception was sometimes thrown by the CountAsync too, so it's not limited to the SearchAsync method**.

The next screenshot is a minimal test that reproduces the behaviour reported. The exception is thrown at the second or third awaited task.
image

### PR DESCRIPTION
**Pull request by flobernd**
Fixes a race condition in the product check.

Closes #8268

### CODE REVIEW

File: src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs [hunk 1]
```diff
@@ -82,17 +92,18 @@ internal ElasticsearchClient(ITransport<IElasticsearchClientSettings> transport)
 	public Serializer SourceSerializer => _transport.Configuration.SourceSerializer;
 	public ITransport<IElasticsearchClientSettings> Transport => _transport;
 
-	private ProductCheckStatus _productCheckStatus;
+	private partial void SetupNamespaces();
+
+	private volatile int _productCheckStatus;
```
Comment (Mpdreamz) [inline src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs:97]: If all reads and write happen through `Interlocked we should be able to get rid of this volotile keyword.

File: src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs [hunk 2]
```diff
@@ -133,48 +144,116 @@ private enum ProductCheckStatus
...
 
-			if (_productCheckStatus == ProductCheckStatus.Failed)
+			var productCheckSucceeded = response.ApiCallDetails.TryGetHeader("x-elastic-product", out var values) &&
+										values.Single().Equals("Elasticsearch", StringComparison.Ordinal);
+
+			_productCheckStatus = productCheckSucceeded
+				? (int)ProductCheckStatus.Succeeded
+				: (int)ProductCheckStatus.Failed;
+
+			if (_productCheckStatus == (int)ProductCheckStatus.Failed)
```
Comment (Mpdreamz) [inline src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs:245]:
```suggestion
			var productCheckStatus = productCheckSucceeded
				? (int)ProductCheckStatus.Succeeded
				: (int)ProductCheckStatus.Failed;
			Interlockd.Exchange(ref _productCheckStatus, productCheckStatus);
			if (productCheckStatus == (int)ProductCheckStatus.Failed)
```

File: src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs [hunk 3]
```diff
@@ -133,48 +144,116 @@ private enum ProductCheckStatus
...
+				response = _transport
+					.Request<TResponse>(request.HttpMethod, resolvedUrl, postData, request.RequestParameters, in openTelemetryData);
+			}
 
-			PostRequestProductCheck<TRequest, TResponse>(request, response);
+			// Evaluate product check result
 
-			if (_productCheckStatus == ProductCheckStatus.Failed)
+			var productCheckSucceeded = response.ApiCallDetails.TryGetHeader("x-elastic-product", out var values) &&
+										values.Single().Equals("Elasticsearch", StringComparison.Ordinal);
```
Comment (Mpdreamz) [inline src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs:239]: Do we want this to throw in `Single()` maybe `FirstOrDefault(...) != null` ?

File: src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs [hunk 4]
```diff
@@ -82,17 +92,18 @@ internal ElasticsearchClient(ITransport<IElasticsearchClientSettings> transport)
...
+
+	private volatile int _productCheckStatus;
 
 	private enum ProductCheckStatus
 	{
-		NotChecked,
-		Succeeded,
-		Failed
+		NotChecked = 0,
+		InProgress = 1,
```
Comment (Mpdreamz) [inline src/Elastic.Clients.Elasticsearch.Shared/Client/ElasticsearchClient.cs:102]: :+1:

**Review by Mpdreamz — CHANGES_REQUESTED** [review]

**Review by Mpdreamz — APPROVED** [review]
LGTM!

Not sure if the 

```
The type or namespace name 'Experimental' does not exist in the namespace 'Elastic.Clients.Elasticsearch' (are you missing an assembly reference?)
``` 

are related.

**Comment by flobernd** [comment]
The tests are not fully updated after the breaking 8.13 release. Have to spend some time on that soon, to have reliable CI builds again.

**Pull request merged**